### PR TITLE
Conversion: Add Expanded VideoPress nudge

### LINKED ISSUE #7648: Conversion: Add Expanded VideoPress Nudge
Breaking up #7330

We should expand the nudge for VideoPress to be more informative. It should look something more like the Advanced SEO nudge in #7540 and be self contained. When you hit upgrade it will add the plan directly to your cart.

screen shot 2016-08-24 at 12 38 24

### PR DESCRIPTION
**Pull request by vindl**
Expands VideoPress nudge in order to make it more informative. 

Closes  #7648 

**Before:**

videobefore

**After:** 

videoafter

- Added icons for Premium and Personal plan to `UpgradeNudgeExpanded` component.
- Fixed minor issue with plan icon display in `UpgradeNudgeExpanded` component - background `no-repeat` property was not applied properly, and repeated icon was visible in the right part of the circle. 

iconbug

Test live: https://calypso.live/?branch=add/expanded-videopress-nudge

### CODE REVIEW

**Comment by vindl** [comment]
cc @roundhill @rodrigoi @dmsnell

**Comment by vindl** [comment]
@apeatling I've also added this on Audio tab. Could you please suggest some benefits that would be more appropriate in that case? 

Currently, the text is the same as in Video tab:
audioafter

File: client/my-sites/media-library/list-plan-promo.jsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { localize } from 'i18n-calypso';
+import VideoPressUpgradeNudge from 'blocks/upgrade-nudge-expanded';
+import { PLAN_PREMIUM, FEATURE_VIDEO_UPLOADS } from 'lib/plans/constants';
+
+export const MediaLibraryListPlanPromo = ( { translate } ) => (
+	<div className="media-library__videopress-nudge-container">
+		<VideoPressUpgradeNudge
+			plan={ PLAN_PREMIUM }
+			title={ translate( 'Upgrade to a Premium Plan and Enable VideoPress' ) }
+			subtitle={ translate( 'By upgrading to a Premium Plan you\'ll enable VideoPress support on your site.' ) }
```
Comment (dmsnell) [inline client/my-sites/media-library/list-plan-promo.jsx]: Might consider using template literals here so we don't need to escape the single quote…

```js
translate( `By upgrading…you'll…your site.` )
```

File: client/my-sites/media-library/list-plan-promo.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import identity from 'lodash/identity';
```
Comment (dmsnell) [inline client/my-sites/media-library/list-plan-promo.jsx]: Not sure if this is better or not, but for consistency's sake we have been importing directly from `lodash`

```js
import { identity } from 'lodash';
```

File: client/my-sites/media-library/list-plan-promo.jsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		<VideoPressUpgradeNudge
+			plan={ PLAN_PREMIUM }
+			title={ translate( 'Upgrade to a Premium Plan and Enable VideoPress' ) }
+			subtitle={ translate( 'By upgrading to a Premium Plan you\'ll enable VideoPress support on your site.' ) }
+			highlightedFeature={ FEATURE_VIDEO_UPLOADS }
+			eventName="calypso_video_uploads_upgrade_nudge_impression"
+			benefits={ [
+				translate( 'Upload videos to your site with an interface designed specifically for WordPress.' ),
+				translate( 'Present videos using a lightweight and responsive player that is ad-free and unbranded.' ),
+				translate( 'See where your videos have been shared as well as stats for individual and overall video plays.' )
```
Comment (dmsnell) [inline client/my-sites/media-library/list-plan-promo.jsx]: As strange as it may seem, I find these sentences ambiguous. We're presenting them as a way of saying, "Look what you could get!" but they also read something like, "Click here to upload videos to your site…"

There's probably some value in noting that to the translators because it could make for some interesting _faux pas_. I guess technically this is a proper place to add in a `context` value since the _meaning_ of the sentence is context-dependent, but I also recently learned about `comment` from @deBhal.

To summarize, I feel like we could benefit here by adding in a translation `context` like `'perk for upgrading wpcom plan'`

File: client/my-sites/media-library/list-plan-promo.jsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Internal dependencies
+ */
+import { localize } from 'i18n-calypso';
+import VideoPressUpgradeNudge from 'blocks/upgrade-nudge-expanded';
+import { PLAN_PREMIUM, FEATURE_VIDEO_UPLOADS } from 'lib/plans/constants';
+
+export const MediaLibraryListPlanPromo = ( { translate } ) => (
+	<div className="media-library__videopress-nudge-container">
+		<VideoPressUpgradeNudge
```
Comment (dmsnell) [inline client/my-sites/media-library/list-plan-promo.jsx]: In another PR I saw this renaming of the upgrade nudge to the component at hand and I'm unsure how I feel about it. It's not a VideoPress component - it's a generic nudge component that we are specializing through passing props. At this point I don't have strong feelings about it, but am definitely uneasy about not calling it something like `UpgradeNudgeExpanded`

Thoughts @artpi ?

cc: @rodrigoi didn't you have something like this recently too with this component?

Comment (artpi) [inline client/my-sites/media-library/list-plan-promo.jsx]: I have no concerns honestly :)

**Comment by artpi** [comment]
Should this be a part of #7853 test? I'd like that.

File: client/my-sites/media-library/list-plan-promo.jsx (hunk 3 above)
Comment (deBhal) [inline client/my-sites/media-library/list-plan-promo.jsx]: In instances like this, adding a `comment` is the way to go.

The acid test for a context is "Will this string need to be translated twice?".  If you're talking about a "table" on the screen as well as a "table" with a bowl of fruit on it, that's going to need a context.

If your string is "Put the fruit on the table", the string itself disambiguates so you don't need the context.   You probably don't need a `comment`, but it can't hurt to add one.

That's the situation here - we can help out the translator by disambiguating with a comment, but there's no real danger of having that same string having a different meaning.

In general: err on the side of including a `comment` and excluding a `context`.

File: client/my-sites/media-library/list-plan-promo.jsx (hunk 1 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-promo.jsx]: Added.

File: client/my-sites/media-library/list-plan-promo.jsx (hunk 2 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-promo.jsx]: Updated.

File: client/my-sites/media-library/list-plan-promo.jsx (hunk 3 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-promo.jsx]: Thanks for the clarification! I've added comments as suggested.

File: client/my-sites/media-library/list-plan-promo.jsx (hunk 4 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-promo.jsx]: I've seen this renaming in other expanded nudge PRs so I continued that practice here for the sake of consistency, but it was strange to me too when I first noticed it. I'm renaming it to `UpgradeNudgeExpanded` here.

**Comment by vindl** [comment]
@artpi Added this nudge to #7853 test.

**Comment by apeatling** [comment]
@vindl Do you still need audio wording?

**Comment by vindl** [comment]
@apeatling Yes, I still need it.

**Comment by apeatling** [comment]
"Upgrade to the Premium Plan and Enable Audio Uploads"

By upgrading to the Premium plan you'll enable audio upload support on your site.

* Add support for podcasting to your site, including a built-in audio player.
* Upload audio files that use any major audio file format.
* Get more than 4 times the storage space for all of your media files.

**Comment by rralian** [comment]
@vindl Can you please change the target of this PR to point to #7853 and rebase it? That way we can launch them together.

**Comment by vindl** [comment]
@rralian rebased and pointed to #7853. 

Also added specific text for audio uploads as @apeatling suggested.

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+import { identity } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { localize } from 'i18n-calypso';
```
Comment (gwwar) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: This is an external dep

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import UpgradeNudgeExpanded from 'blocks/upgrade-nudge-expanded';
+import { PLAN_PREMIUM, FEATURE_VIDEO_UPLOADS, FEATURE_AUDIO_UPLOADS } from 'lib/plans/constants';
+
+export const MediaLibraryUpgradeNudge = ( { translate, filter } ) => {
+	let title = translate( 'Upgrade to a Premium Plan and Enable VideoPress' );
+	let subtitle = translate( `By upgrading to a Premium Plan you'll enable VideoPress support on your site.` );
```
Comment (gwwar) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: use `const` over `let` if you're not reassigning the reference.

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			'Present videos using a lightweight and responsive player that is ad-free and unbranded.',
+			{ comment: 'Perk for upgrading to premium plan' }
+		),
+		translate(
+			'See where your videos have been shared as well as stats for individual and overall video plays.',
+			{ comment: 'Perk for upgrading to premium plan' }
+		)
+	];
+
+	if ( 'audio' === filter ) {
```
Comment (gwwar) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: The length of this stateless function make it difficult to read. Can you move some of this to helper functions?

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx (hunk 6 above)
Comment (gwwar) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: Is your eslint setup in your editor? We should use single strings here. 

We can also run `make eslint-branch` locally to lint the related files

File: client/my-sites/media-library/list.jsx [hunk 8]
```diff
@@ -198,9 +201,11 @@ module.exports = React.createClass( {
 		var onFetchNextPage;
 
 		if ( this.props.filterRequiresUpgrade ) {
-			return (
-				<ListPlanPromo site={ this.props.site } filter={ this.props.filter } />
-			);
+			if( 'expanded' === abtest( 'expandedNudge' ) ) {
```
Comment (gwwar) [inline client/my-sites/media-library/list.jsx]: We don't need to do yoda conditions in javascript

**Comment by rralian** [comment]
I keep running into issues running this branch. It will work initially, but then after loading it a few times, it will stop rendering the nudge. For some reason it will just return null for the nudge.

null nudge

I've tested and when this happens the `<ExpandedUpgradeNudge` fails [on this line](https://github.com/Automattic/wp-calypso/blob/b6cc69b036e39953151c6a3a83c45c1d5e15903a/client/blocks/upgrade-nudge-expanded/index.jsx#L53) because `this.props.currentPlan` is null. I'm not sure how/why this is happening. I'm testing with my test site bobinprogress.wordpress.com. I've also recreated with my test site asdf257.wordpress.com.

File: client/blocks/upgrade-nudge-expanded/style.scss [hunk 9]
```diff
@@ -44,13 +44,21 @@
 	}
 
 	.upgrade-nudge-expanded__title-plan-icon {
-		width: 54px;
-		height: 54px;
+		width: 55px;
```
Comment (gwwar) [inline client/blocks/upgrade-nudge-expanded/style.scss]: We should update the nudge to use a plan icon component instead:
https://github.com/Automattic/wp-calypso/tree/master/client/components/plans/plan-icon
e.g:
```javascript
<PlanIcon plan={ PLAN_BUSINESS } />
```

File: client/lib/plans/constants.js [hunk 10]
```diff
@@ -371,6 +378,16 @@ export const featuresList = {
...
 	},
 
+	[ FEATURE_AUDIO_UPLOADS ]: {
+		getSlug: () => FEATURE_AUDIO_UPLOADS,
+		getTitle: () => i18n.translate( 'Audio Upload Support' ),
+		getDescription: () => i18n.translate(
+			'The easiest way to upload audio files that use any major audio file format. '
+		),
+		getStoreSlug: () => 'videopress',
+		plans: allPaidPlans
```
Comment (gwwar) [inline client/lib/plans/constants.js]: We don't get video for the personal plan.

File: client/my-sites/media-library/list.jsx [hunk 11]
```diff
@@ -20,6 +20,9 @@ var MediaActions = require( 'lib/media/actions' ),
 	InfiniteList = require( 'components/infinite-list' ),
 	user = require( 'lib/user' )();
 
+import ListPlanUpgradeNudge from './list-plan-upgrade-nudge';
+import { abtest } from 'lib/abtest';
```
Comment (gwwar) [inline client/my-sites/media-library/list.jsx]: There aren't too many imports, so let's tidy the rest of the requires.

File: client/my-sites/media-library/list.jsx [hunk 12]
```diff
@@ -198,9 +201,11 @@ module.exports = React.createClass( {
 		var onFetchNextPage;
 
 		if ( this.props.filterRequiresUpgrade ) {
-			return (
-				<ListPlanPromo site={ this.props.site } filter={ this.props.filter } />
-			);
+			if( 'expanded' === abtest( 'expandedNudge' ) ) {
+				return <ListPlanUpgradeNudge filter={ this.props.filter } />;
+			} else {
```
Comment (gwwar) [inline client/my-sites/media-library/list.jsx]: The else is optional here, and will cause an eslint warning.

**Comment by rralian** [comment]
The issue I am running into seems to be a problem with the `getCurrentPlan()` selector. It calls `getPlansBySiteId` and expects to have data in either `state.sites.plans` or `initialSiteState` and neither of these are the case for me when I run into this issue. The problem is on this line.

https://github.com/Automattic/wp-calypso/blob/b6cc69b036e39953151c6a3a83c45c1d5e15903a/client/state/sites/plans/selectors.js#L33

I think perhaps this is an issue where the `UpgradeNudgeExpanded` component needs an additional query. /cc @artpi

**Comment by gwwar** [comment]
The component only queries the /plans endpoint rather than the site specific one, so I would suggest using: `getSitePlan` selector that uses plan information off of the site object directly.

https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/selectors.js#L452

File: client/my-sites/media-library/style.scss [hunk 13]
```diff
@@ -115,3 +115,27 @@
...
+	position: absolute;
+	left: 0;
+	right: 0;
+	box-shadow: none;
+	margin: 0;
+
+	@include breakpoint( ">960px") {
+		top: 50%;
+		transform: translateY(-50%);
+	}
```
Comment (rralian) [inline client/my-sites/media-library/style.scss]: This is just my opinion, but I really don't care for the vertical centering we're trying to do in wider breakpoints. I'd rather just see it tied to the top of the page with some padding.

File: client/my-sites/media-library/style.scss [hunk 14]
```diff
@@ -115,3 +115,27 @@
 .media-library__list-item.is-small .media-library__list-item-file-extension {
 	display: none;
 }
+
+.media-library__videopress-nudge-container {
+	display: block;
+	overflow-y: auto;
+	position: absolute;
```
Comment (rralian) [inline client/my-sites/media-library/style.scss]: Is the absolute positioning really necessary here? Seems like we're just adding some padding/margin to the top. It's best to avoid absolute position unless there's some reason that we need it here.

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx (hunk 5 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: Updated.

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx (hunk 7 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: Moved to helper functions as suggested.

File: client/my-sites/media-library/list-plan-upgrade-nudge.jsx (hunk 6 above)
Comment (vindl) [inline client/my-sites/media-library/list-plan-upgrade-nudge.jsx]: No, I haven't set it up yet. Reverted to using single strings here.

File: client/my-sites/media-library/list.jsx (hunk 8 above)
Comment (vindl) [inline client/my-sites/media-library/list.jsx]: I am aware of that, I thought I've read somewhere in our coding guidelines that they are optional. Removed them from this PR.

File: client/lib/plans/constants.js (hunk 10 above)
Comment (vindl) [inline client/lib/plans/constants.js]: Updated for video and audio uploads.

File: client/my-sites/media-library/style.scss (hunk 14 above)
Comment (vindl) [inline client/my-sites/media-library/style.scss]: As far as I remember I've ran into some issues when trying to position it, because of absolute positioning of some other parts of media container. It might not be necessary and I'll try to avoid it.

**Comment by artpi** [comment]
I cherry-picked these commits into #7853 so that we have one PR to merge with the tests.
I also think, we can close this PR and continue discussion in #7853

> I think perhaps this is an issue where the UpgradeNudgeExpanded component needs an additional query

Thanks @rralian for resolving this in #8054 . I merged it and the issue seems resolved

> The component only queries the /plans endpoint rather than the site specific one, so I would suggest using: getSitePlan

That seems to be resolved in #8054.

File: client/my-sites/media-library/list.jsx (hunk 11 above)
Comment (artpi) [inline client/my-sites/media-library/list.jsx]: Please, lets do it in Janitorial one.

File: client/blocks/upgrade-nudge-expanded/style.scss (hunk 9 above)
Comment (artpi) [inline client/blocks/upgrade-nudge-expanded/style.scss]: Fixed in c117696c70474603daa1bc4c6e41254382f11c2a

**Comment by vindl** [comment]
Closing this PR and continuing in #7853 per @artpi's request.
